src/mysql.erl

```diff
@@ -175,32 +181,86 @@ start_link(Options) ->
 query(Conn, Query, Params) when is_list(Params) ->
-    query_call(Conn, {param_query, Query, Params});
+    query_call(Conn, {param_query, Query, Params, undefined, undefined});
+query(Conn, Query, FilterMap) when FilterMap=:=undefined;
+        is_function(FilterMap, 1); is_function(FilterMap, 2) ->
+    query_call(Conn, {query, Query, FilterMap, undefined});
 query(Conn, Query, Timeout) when is_integer(Timeout); Timeout == infinity ->
-    query_call(Conn, {query, Query, Timeout}).
+    query_call(Conn, {query, Query, undefined, Timeout}).
 
-%% @doc Executes a parameterized query with a timeout.
+%% @doc Depending on the 3rd and 4th arguments this function does different
+%% things.
+%%
+%% If either the 3rd argument is a list, a prepared statement is created,
+%% executed and then cached for a certain time. If the same query is executed
+%% again when it is already cached, it does not need to be prepared again.
+%% The minimum time the prepared statement is cached can be specified using the
+%% option `{query_cache_time, Milliseconds}' to start_link/1.
+%% Conversely, if the 3rd argument is not a list, a plain query is executed.
+%%
+%% If the 4th argument is a timeout, the query is executed with that timeout.
+%% Conversely, if the 4th argument is not a timeout, the timeout given in
+%% start_link/1 is used.
+%%
+%% If either the 3rd or 4th argument is a function instead of a list or
+%% timeout, respectively, it is used to filter/map the result rows.
+%%
+%% The return value is the same as for query/2.
+-spec query(Conn, Query, Params, Timeout) -> Result
+        when Conn :: connection(),
+             Query :: iodata(),
+             Timeout :: timeout(),
+             Params :: [term()],
+             Result :: query_result();
+    (Conn, Query, FilterMap, Timeout) -> Result
+        when Conn :: connection(),
+             Query :: iodata(),
+             Timeout :: timeout(),
+             FilterMap :: query_filtermap(),
+             Result :: query_result();
+    (Conn, Query, Params, FilterMap) -> Result
+        when Conn :: connection(),
+             Query :: iodata(),
+             Params :: [term()],
+             FilterMap :: query_filtermap(),
+             Result :: query_result().
+query(Conn, Query, Params, Timeout) when is_list(Params) andalso (is_integer(Timeout) orelse Timeout=:=infinity) ->
+    query_call(Conn, {param_query, Query, Params, undefined, Timeout});
+query(Conn, Query, FilterMap, Timeout) when (is_function(FilterMap, 1) orelse is_function(FilterMap, 2)) andalso (is_integer(Timeout) orelse Timeout=:=infinity) ->
+    query_call(Conn, {query, Query, FilterMap, Timeout});
+query(Conn, Query, Params, FilterMap) when is_list(Params) andalso (is_function(FilterMap, 1) orelse is_function(FilterMap, 2)) ->
+    query_call(Conn, {param_query, Query, Params, FilterMap, undefined}).
+
+%% @doc Executes a parameterized query with a timeout and applies a filter/map
+%% function to the result rows..
```

Comment:
Yep....

I think we should put the full descriptions here and just put "see query/5" at the other other query/2,3,4 functions.

Reply:
Hm, for that to be consistent, it should be possible to have a `query/5` call that is equivalent to a `query/2` call (in other words, `query/2` should be able to map to `query/5`), which is in fact _not_ possible, as `query/5` is always a parameterized query, while `query/2` is always a plain query.  It is possible to have equivalent _outcomes_ by giving a plain query and an empty list to `query/5`, but it would be handled different internally. Not sure it matters to end users, but it may make a difference to people working on `mysql-otp`.

Maybe it would be clearer to say something like "for an explanation of the _arguments_, see `query/5`" or something?

Reply:
That's true.

So, what if we allow `no_params`, `no_filtermap_fun` and `default_timeout` in `query/5`? Then all other variants can simply call this one.

Then, we can write in the docs "Rather than passing `no_params`, `no_filtermap_fun` and `default_timeout`, consider using `query/2,3,4` instead" as a suggestion, while still allowing fun logic like `mysql:query(Pid, Query, if X -> fun(X) -> {true, Y} end; true -> no_filtermap_fun)`.

As for the style with optional arguments, it is used extensively in OTP, e.g. `proplists:get_value/2,3`, which is fine as long as you don't want to omit some param in the middle. Originally, the only optional argument was Params, which was fine then, but it was extended later with more optional params.....

Reply:
> That's true.
> 
> So, what if we allow `no_params`, `no_filtermap_fun` and `default_timeout` in `query/5`? Then all other variants can simply call this one.

Ok, that's good :) I'll go that route. It may mean changes to other functions (like `prepare`, I guess) in order to make them work in a similar fashion, which is actually unrelated to this PR.

> As for the style with optional arguments, it is used extensively in OTP, e.g. `proplists:get_value/2,3`, which is fine as long as you don't want to omit some param in the middle. Originally, the only optional argument was Params, which was fine then, but it was extended later with more optional params.....

Yes, I know. As of lately, in my impression, they tend to avoid the "leave out stuff in the middle" variants that are possible if argument types make them discernible and are tending more to just using separate functions, or option maps (or proplists, before the advent of maps) for this purpose. I think if it ever came to a new/alternative API for `mysql-otp`, option maps (or rather proplists, since you are also targeting OTP releases that don't have them, like R16) would be the best way to do it. Something like
```erlang
-spec query2(Conn :: pid(), Query :: iodata(), Options :: [{Key :: atom(), Value :: term()}]) -> query_result().
...
-spec param_query2(Conn :: pid(), Query :: iodata(), Params :: [term()], Options :: [{Key :: atom(), Value :: term()}]) -> query_result().
...
```
where timeout, filtermap fun etc would go into the options.

However, this also is something to be discussed separately. Which reminds me, if you ever get to #erlang on freenode (IRC), ping me, I often hang out there during the day, nick is the same as my github handle ;)

Reply:
If we ever change this interface, we can make a version 2.0 and drop support for those old OTP versions. Maps are available from OTP 19.

I'll find you on IRC!

Reply:
IIRC, since OTP 17, even :)